File: app/src/main/java/dev/ragnarok/fenrir/model/FunctionSource.java
package dev.ragnarok.fenrir.model;

import android.content.Context;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

public class FunctionSource {
    private final Text title;
    private final FunctionSourceCall call;
    private final @DrawableRes
    int icon;

    public FunctionSource(String title, @DrawableRes int icon, @NonNull FunctionSourceCall call) {
        this.title = new Text(title);
        this.call = call;
        this.icon = icon;
    }

    public FunctionSource(@StringRes int title, @DrawableRes int icon, @NonNull FunctionSourceCall call) {
        this.title = new Text(title);
        this.call = call;
        this.icon = icon;
    }

    public String getTitle(Context context) {
        return title.getText(context);
    }

    public void Do() {
        call.call();
    }

    public @DrawableRes
    int getIcon() {
        return icon;
    }

    public interface FunctionSourceCall {
        void call();
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/model/FunctionSource.kt
package dev.ragnarok.fenrir.model

import android.content.Context
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes

class FunctionSource {
    private val title: Text
    private val call: () -> Unit

    @DrawableRes
    private val icon: Int

    constructor(title: String?, @DrawableRes icon: Int, call: () -> Unit) {
        this.title = Text(title)
        this.call = call
        this.icon = icon

    }

    constructor(@StringRes title: Int, @DrawableRes icon: Int, call: () -> Unit) {
        this.title = Text(title)
        this.call = call
        this.icon = icon
    }

    fun getTitle(context: Context?): String? {
        return if (context == null) {
            null
        } else title.getText(context)
    }

    fun Do() {
        call.invoke()
    }

    @DrawableRes
    fun getIcon(): Int {
        return icon
    }
}